### PR TITLE
Bug/ICS-22 cleanup duplicate namespaces

### PR DESCRIPTION
**Pull request by ahmedolaibrahim**
- [x]  CleanUp Duplicate ENS namespaces
- [x] Prevent Duplicates 

To Test The Duplicates Removal

- Create a dump of staging DB
- Create a fresh DB locally and import dump from staging
- Run migration Command

### CODE REVIEW

File: src/migrations/1630744507207-RemoveDuplicates.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+const deleteDuplicateApps = async (queryRunner: QueryRunner) => {
+  await queryRunner.query(
+    `DELETE FROM application app
+            WHERE  EXISTS (
+            SELECT FROM application
+            WHERE  namespace = app.namespace
+            AND    id < app.id
+        );`,
+  );
+};
```
Comment (ahmedolaibrahim) [inline src/migrations/1630744507207-RemoveDuplicates.ts:23]: Duplicates Apps found in the database all turned out to have the same `parentOrgId` of the first OrgId created before other duplicate orgs hence it makes sense for us to remove duplicates except the first one created.

File: src/migrations/1630744507207-RemoveDuplicates.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+const deleteDuplicateRoles = async (queryRunner: QueryRunner) => {
+  await queryRunner.query(
+    `DELETE FROM role r
+            WHERE  EXISTS (
+            SELECT FROM role
+            WHERE  namespace = r.namespace
+            AND    id < r.id
+        );`,
+  );
+};
```
Comment (ahmedolaibrahim) [inline src/migrations/1630744507207-RemoveDuplicates.ts:34]: same reason as above

File: src/migrations/1630744507207-RemoveDuplicates.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+const deleteDuplicateOrgs = async (queryRunner: QueryRunner) => {
+  await queryRunner.query(
+    `DELETE FROM organization org
```
Comment (ahmedolaibrahim) [inline src/migrations/1630744507207-RemoveDuplicates.ts:5]: removes all duplicate except the first created org.

File: src/modules/organization/organization.service.ts [hunk 4]
```diff
@@ -97,8 +97,18 @@ export class OrganizationService {
...
-    return Boolean(await this.getByNamespace(namespace));
+  public async exists(
+    namespace: string,
+    parentOrg?: Organization,
+  ): Promise<boolean> {
+    const whereObjectClause: any = { namespace };
+    if (parentOrg) {
+      whereObjectClause.parentOrg = parentOrg;
+    }
+    const appExists = await this.orgRepository.findOne({
```
Comment (JGiter) [inline src/modules/organization/organization.service.ts:108]:
```suggestion
    const orgExists = await this.orgRepository.findOne({
```

File: src/modules/role/role.service.ts [hunk 5]
```diff
@@ -46,8 +48,26 @@ export class RoleService {
    * return true if role with given namespace exists
    * @param namespace
    */
-  public async exists(namespace: string) {
-    return Boolean(await this.getByNamespace(namespace));
+  public async exists(
```
Comment (JGiter) [inline src/modules/role/role.service.ts:51]: This signature does not allow to check only parent organization

File: src/modules/role/role.fixture.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      metadata: {},
+    };
+
+    const role = Role.create({
+      name,
+      namespace: `${name}.roles.testapp.apps.testApp.iam.ewc`,
+      owner: '0x7dD4cF86e6f143300C4550220c4eD66690a655fc',
+      definition,
+      parentOrg: organization,
+      parentApp: application,
```
Comment (JGiter) [inline src/modules/role/role.fixture.ts:41]: Role has only one parent which is app or org

**Review by JGiter — CHANGES_REQUESTED** [review]

File: src/modules/application/application.service.spec.ts [hunk 7]
```diff
@@ -1,19 +1,127 @@
...
-  it('should be defined', () => {
-    expect(service).toBeDefined();
+  afterEach(async () => {
+    await queryRunner.rollbackTransaction();
+    module.close();
+  });
+
+  it('create() it should not save an application with existing namespace and parentOrg ', async () => {
+    const testApp = chance.pickone(applications);
+    const organization = chance.pickone(organizations);
```
Comment (jrhender) [inline src/modules/application/application.service.spec.ts:85]: Why the use of `chance` here? I tend to think tests should not have randomness so that they are repeatable, but maybe I don't understand the rationale here

File: src/modules/application/application.service.spec.ts [hunk 8]
```diff
@@ -1,19 +1,127 @@
...
+      owner,
+      definition,
+      parentOrg: organization.namespace,
+    });
+
+    expect(MockLogger.debug).toHaveBeenCalledWith(
+      expect.stringContaining(
+        `Not able to create application: ${namespace} with same parent organization ${organization.namespace} already exists`,
+      ),
+    );
```
Comment (jrhender) [inline src/modules/application/application.service.spec.ts:102]: I think testing the logger is a find approach, but maybe it would be even better to query the db and ensure that there aren't duplicates using the `service`?

File: src/modules/application/application.service.ts [hunk 9]
```diff
@@ -60,8 +61,18 @@ export class ApplicationService {
...
+    parentOrg?: Organization,
+  ): Promise<boolean> {
+    const whereObjectClause: any = { namespace };
+    if (parentOrg) {
+      whereObjectClause.parentOrg = parentOrg;
+    }
+    const appExists = await this.applicationRepository.findOne({
+      where: whereObjectClause,
+    });
+    return Boolean(appExists);
```
Comment (jrhender) [inline src/modules/application/application.service.ts:75]: Why this change? Is it fixing a bug or just adding the ability to also have a check for the `parentOrg`. I would think that the `parentOrg` is implied in the namespace already. Like if a namespace like `finance.apps.ahmed.iam.ewc` is passed a `namespace`, aren't we already checking that the parent is `ahmed.iam.ewc` implicitly 🤔 ?

File: src/modules/application/application.service.ts [hunk 10]
```diff
@@ -60,8 +61,18 @@ export class ApplicationService {
    * return true if App with given namespace exists
    * @param namespace
    */
```
Comment (jrhender) [inline src/modules/application/application.service.ts:63]: Missing documentation for new `parentOrg` param

File: src/modules/role/role.service.ts [hunk 11]
```diff
@@ -46,8 +48,26 @@ export class RoleService {
    * return true if role with given namespace exists
    * @param namespace
    */
```
Comment (jrhender) [inline src/modules/role/role.service.ts:50]: Missing documentation

File: src/modules/organization/organization.service.ts [hunk 12]
```diff
@@ -97,8 +97,18 @@ export class OrganizationService {
    * return true if Org with given namespace exists
    * @param namespace
    */
```
Comment (jrhender) [inline src/modules/organization/organization.service.ts:99]: Missing documentation

File: src/modules/role/role.fixture.ts (hunk 6 above)
Comment (jrhender) [reply src/modules/role/role.fixture.ts:41]: I agree with Dmitry. This should be enforced in `Role.create`, I believe

**Review by jrhender — CHANGES_REQUESTED** [review]
Some questions and missing documentation

File: src/modules/application/application.service.spec.ts (hunk 7 above)
Comment (ahmedolaibrahim) [reply src/modules/application/application.service.spec.ts:85]: Yeah I agree with you that tests should not have randomness but in this case, using chance to pick one of an already created app or org doesn't affect the outcome of the tests because what is selected is what is going to be saved at the end and that is what we are asserting for.

File: src/modules/application/application.service.spec.ts (hunk 8 above)
Comment (ahmedolaibrahim) [reply src/modules/application/application.service.spec.ts:102]: Not sure that would be useful in this case though, remember we are trying to test if we are able to save or not  and the Logger is only called when we are not able to save it. querying the DB would be an additional call we don't actually need.

File: src/modules/application/application.service.ts (hunk 9 above)
Comment (ahmedolaibrahim) [reply src/modules/application/application.service.ts:75]: While cleaning up duplicates, I discovered that, There are cases where we have duplicates with the same `namespace` and `parentOrg` so this is more like a check for if that namespace with that parentOrg already exists.

File: src/modules/role/role.service.ts (hunk 5 above)
Comment (ahmedolaibrahim) [reply src/modules/role/role.service.ts:51]: This exists method is for the `roles`  i don't think there's a reason for us to be checking for only a parentOrg here, we can either check only for a role namespace or role namespace belonging to a parentOrg. we can check for only parent org namespace in the exists method for organisation.

Comment (JGiter) [reply src/modules/role/role.service.ts:53]: I mean that optional arguments can be omitted, but `parentApp` can't be omitted, when you want to filter only by `parentOrg`

File: src/modules/application/application.service.spec.ts (hunk 8 above)
Comment (jrhender) [reply src/modules/application/application.service.spec.ts:102]: Hmm, I was suggesting querying the DB in the test to ensure that we don't have a duplicate because that ultimately what we care about. Is that was you're talking about too?
My rationale is that, I know it's not currently doing this, but it would be possible to log that message but still somehow create a duplicate in the DB, in which case the test would be misleading IMO.

File: src/modules/application/application.service.ts (hunk 9 above)
Comment (jrhender) [reply src/modules/application/application.service.ts:76]: Hmm, okay, I'm still not fully understanding when someone would use the `parentOrg` param. What I mean is that, if I'm querying for a role like `engineer.roles.ahmed.iam.ewc`, the parent org can only be `ahmed.iam.ewc`. So if someone were to make a query like `exists(engineer.roles.ahmed.iam.ewc, john.iam.ewc)`, that is obviously going to be false. Am I misunderstanding something here?

Comment (ahmedolaibrahim) [reply src/modules/application/application.service.ts:76]: Sorry for the misunderstanding. you're right. Dmitry clarified this to me this morning.

File: src/modules/application/application.service.spec.ts (hunk 8 above)
Comment (ahmedolaibrahim) [reply src/modules/application/application.service.spec.ts:102]: okay make sense.

**Comment by ahmedolaibrahim** [comment]
@jrhender @JGiter Closing this in favour of [this](https://github.com/energywebfoundation/iam-cache-server/pull/158). the other PR contains changes requested here and  namespace deletion sync.
